### ISSUE TITLE
implement mint with permission

### ISSUE THREAD

**Issue by TimDaub** [rugpullindex/ERC4973#7]
- Generally, implement `function mintWithPermission(address from, address to, uint256 tokenId, string tokenURI, uint8 v, bytes32 r, bytes32 s)` where the following is signed as an EIP721: `sign(hash(from.address, to, tokenURI), from.privateKey))`
- Rely on "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol"
- Consider making the implementation maximally similar to
  - https://eips.ethereum.org/EIPS/eip-2612#specification
  - https://eips.ethereum.org/EIPS/eip-4494

1. Send a PR to timdaub/EIPS that introduces the above defined function signature as an OPTIONAL extension in the "Specification" section
	- In the "Rationale" section, write a one-liner why `mintWithPermission` is useful. It's useful because both the `from` and `to` arguments of an ABT are authenticated by Ethereum and hence mutually agreed consent between two addresses can be expressed.
	- In "Specification" cannonically outline what MUST be hashed and signed: from, to, tokenURI. Clarify that tokenURI must be content-addressable. Use language from https://eips.ethereum.org/EIPS/eip-5114 about content-addressing which is the best currently.
	- In "Rationale" or "Security Considerations" make a statement that notes that if a "tokenURI" isn't content-addressable, mintWithPermission cannot be achieved as the underlying metadata of "tokenURI" could change after signing.
	- Have the PR reviewed by me
2. In the reference implementation github repo: https://github.com/rugpullindex/ERC4973, send a PR with a implementation of the above defined function
  - Write unit tests that show that everything works correctly
  - Write unit tests that show that in case of valid/malicious inputs, nothing breaks and everything works safely
  - Create a PR and have it reviewed by me. Once it's merged, I'll take care of both publishing the reference implementation code and change to the EIP document.

**Comment by rahulrumalla** [comment]
I can take this on

**Comment by rahulrumalla** [comment]
Heya. I'm actively working on this.

I'm wondering if the typehash needs to be `Approval(issuer,recipient,token_id,token_uri)`.  Any particular reason you wanted to not include token_id in there in your spec above?

**Comment by rahulrumalla** [comment]
Good point. Excuse the confusion from my language above - have been using camel case and 'Permit' in actual implementation so far. Will issue a draft PR soon so u can see the WIP

**Comment by rahulrumalla** [comment]
> > Good point. Excuse the confusion from my language above - have been using camel case and 'Permit' in actual implementation so far. Will issue a draft PR soon so u can see the WIP
> 
> no worries! And looking forward to seeing your PR!

Draft PR is ready for review.

**Comment by rahulrumalla** [comment]
>make a statement that notes that if a "tokenURI" isn't content-addressable, mintWithPermission cannot be achieved as the underlying metadata of "tokenURI" could change after signing.

Looking into Content Addressability. I think for an NFT standard to truly achieve this, I think we need to be able to set baseURI in constructor and then only take CIDs as parameter in `mint*()` in order to construct the tokenURI in the contract itself, no? 

Were you implying that the responsibility of achieving content addressability in tokenURI is consciously deferred to the caller ?of the contract:?

**Comment by TimDaub** [comment]
> Looking into Content Addressability. I think for an NFT standard to truly achieve this, I think we need to be able to set baseURI in constructor and then only take CIDs as parameter in `mint*()` in order to construct the tokenURI in the contract itself, no?

IMO content-addressing of content in tokenURI cannot be strictly enforced. Like a contract will never be able to tell if the data behind a URI is content-addressed or not. A contract would have to be capable of downloading, hashing and verifying: And it'll not be capable of doing so.

Rather, I think we should clarify in the specification document that: If the metadata resulting from GET'ing the tokenURI can change (because it isn't content-addressed) then it's bad because the authenticity of the `mintWithPermission` function is at risk.

So rather than mandating users to only implement content-addressable tokenURIs, I'd frame it in the text of the EIP as: "The only logical way of using `mintWithPermission` is to use content-addressable hashes as otherwise, you could be signing a URL with content X and then later the underlying content is changed to Y after the fact." The effect of this framing is that anybody understanding the specification would use a content-addressable hash as a tokenURI.
